tests/lsp/test_completion.py

```
@@ -0,0 +1,81 @@
+from pygls.server import LanguageServer
+
+from ..conftest import CALL_TIMEOUT, ClientServer
+
+
+class TestCompletions(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.client_server = ClientServer()
+        cls.client, cls.server = cls.client_server
+
+        @cls.server.feature(
+            COMPLETION,
+            CompletionOptions(trigger_characters=[','], all_commit_characters=[':'], resolve_provider=True)
+        )
+        def f(params: CompletionParams) -> CompletionList:
+            return CompletionList(
+                is_incomplete=False,
+                items=[
+                    CompletionItem(
+                        label='test1',
+                        kind=CompletionItemKind.Method,
+                        preselect=True,
+                    ),
+                ]
+            )
+
+        cls.client_server.start()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.client_server.stop()
+
+    def test_capabilities(self):
+        capabilities = self.server.server_capabilities
+
+        assert capabilities.completion_provider
+        assert capabilities.completion_provider.trigger_characters == [',']
+        assert capabilities.completion_provider.all_commit_characters == [':']
+        assert capabilities.completion_provider.resolve_provider is True
+
+    def test_completions(self):
+        response = self.client.lsp.send_request(
+            COMPLETION,
+            CompletionParams(
+                text_document=TextDocumentIdentifier(uri='file://test.test'),
+                position=Position(line=0, character=0)
+            )
+        ).result(timeout=CALL_TIMEOUT)
+
+        assert response['isIncomplete'] == False
+        assert response['items'][0]['label'] == 'test1'
+        assert response['items'][0]['kind'] == CompletionItemKind.Method
+        assert response['items'][0]['preselect'] == True
+        assert not hasattr(response['items'][0], 'deprecated')
```

Great work on this PR, as an end-user I really look forward to the changes!

Would `'deprecated' not in response['items'][0]` be better here (and in other places where these things are being checked)? I mean `hasattr({'a': 1}, 'a') is False` so this check may be silently passing if the underlying `CompletionItem` is a dictionary (which I would expect).

---

Thanks! Good catch, will fix it.

Have you actually tried out these changes? Did you notice any issues with how it works?

---

No I have not tested it, only reviewed the code to see how it fixes the `null`s. I could try to run it next weekend if you would like some more feedback.

---

I have pushed the fix.

> I could try to run it next weekend if you would like some more feedback.

That would be great. I hope I will find some time to update the docs in the meantime.

---

Did you find time to try it out?